compiler/core/js_exp_make.ml

```diff
@@ -716,224 +716,328 @@ let rec push_negation (e : t) : t option =
-          {expression_desc = Typeof {expression_desc = Var ia}},
-          {expression_desc = Str {txt = "string"}} ) )
-    when Js_op_util.same_vident ia ib ->
-    None
-  | ( Bin
-        ( EqEqEq,
-          {expression_desc = Typeof {expression_desc = Var ia}},
-          {expression_desc = Str {txt = "number"}} ),
-      Bin (NotEqEq, {expression_desc = Var ib}, {expression_desc = Number _}) )
-  | ( Bin (NotEqEq, {expression_desc = Var ib}, {expression_desc = Number _}),
-      Bin
-        ( EqEqEq,
-          {expression_desc = Typeof {expression_desc = Var ia}},
-          {expression_desc = Str {txt = "number"}} ) )
-    when Js_op_util.same_vident ia ib ->
-    None
-  | ( (Bin
-         ( EqEqEq,
-           {expression_desc = Typeof {expression_desc = Var ia}},
-           {expression_desc = Str {txt = "boolean" | "string" | "number"}} ) as
-       typeof),
-      Bin
-        ( NotEqEq,
-          {expression_desc = Var ib},
-          {expression_desc = Bool _ | Null | Undefined _ | Number _ | Str _} ) )
-  | ( Bin
-        ( NotEqEq,
-          {expression_desc = Var ib},
-          {expression_desc = Bool _ | Null | Undefined _} ),
-      (Bin
-         ( EqEqEq,
-           {expression_desc = Typeof {expression_desc = Var ia}},
-           {expression_desc = Str {txt = "boolean" | "string" | "number"}} ) as
-       typeof) )
-    when Js_op_util.same_vident ia ib ->
-    (* Note: cases boolean / Bool _, number / Number _, string / Str _ are handled above *)
-    Some {expression_desc = typeof; comment = None}
-  | ( (Call
-         ( {expression_desc = Str {txt = "Array.isArray"}},
-           [{expression_desc = Var ia}],
-           _ ) as is_array),
-      Bin
-        ( NotEqEq,
-          {expression_desc = Var ib},
-          {expression_desc = Bool _ | Null | Undefined _ | Number _ | Str _} ) )
-  | ( Bin
-        ( NotEqEq,
-          {expression_desc = Var ib},
-          {expression_desc = Bool _ | Null | Undefined _ | Number _ | Str _} ),
-      (Call
-         ( {expression_desc = Str {txt = "Array.isArray"}},
-           [{expression_desc = Var ia}],
-           _ ) as is_array) )
-    when Js_op_util.same_vident ia ib ->
-    Some {expression_desc = is_array; comment = None}
-  | x, y when x = y -> Some e1
-  | ( Bin
-        ( EqEqEq,
-          {expression_desc = Var ia},
-          {expression_desc = Bool _ | Null | Undefined _ | Number _ | Str _} ),
-      Bin
-        ( EqEqEq,
-          {expression_desc = Var ib},
-          {expression_desc = Bool _ | Null | Undefined _ | Number _ | Str _} ) )
-    when Js_op_util.same_vident ia ib ->
-    (* Note: case x = y is handled above *)
-    Some false_
-  | _ -> None
+    Printf.eprintf "%s simplify_and %s %s\n"
+      (String.make (n * 2) ' ')
+      (!string_of_expression e1) (!string_of_expression e2);
+  let res =
+    match (e1.expression_desc, e2.expression_desc) with
+    | Bool false, _ -> Some false_
+    | _, Bool false -> Some false_
```

Comment:
Does this need a check that `e1` doesn't have side effects?

Reply:
Thanks! It does.

Reply:
https://github.com/rescript-lang/rescript-compiler/pull/7151/commits/0d49895dc9f272fdb16ebb704b7a75deec4772d6

Reply:
@yallop nice find! See any more issues of that kind? 👀

Reply:
Yes, there seem to be a few.  To pick some at random:
- in [`eq_null_undefined_boolean`](https://github.com/rescript-lang/rescript-compiler/blob/9b08fe929eb8aa06a0ab58f1731de24315eddedf/compiler/core/js_exp_make.ml#L1710-L1712)
   ```ocaml
  match (a.expression_desc, b.expression_desc) with
  | ( (Null | Undefined _),
      (Bool _ | Number _ | Typeof _ | Fun _ | Array _ | Caml_block _) ) ->
   ```
  the `Typeof` or `Array` or `Caml_block` could have side effects, so can't always be safely dropped
-  in [`int_comp`](https://github.com/rescript-lang/rescript-compiler/blob/9b08fe929eb8aa06a0ab58f1731de24315eddedf/compiler/core/js_exp_make.ml#L1374-L1375)
   ```ocaml
     | Ceq, Optional_block _, Undefined _ | Ceq, Undefined _, Optional_block _ ->
    false_
    ```
   if `Optional_block` can have side effects then it shouldn't be dropped
- in [`simplify_or`](https://github.com/rescript-lang/rescript-compiler/blob/9b08fe929eb8aa06a0ab58f1731de24315eddedf/compiler/core/js_exp_make.ml#L1071)
  ```ocaml
      | _, Bool true -> Some true_
  ```
  the left operand shouldn't be dropped if it has side effects.
  
This is just a selection; there are some similar cases in those functions and others.  I don't have all the development tools set up, so can't easily send a PR, unfortunately.

Reply:
> Yes, there seem to be a few. To pick some at random:
> 
> * in [`eq_null_undefined_boolean`](https://github.com/rescript-lang/rescript-compiler/blob/9b08fe929eb8aa06a0ab58f1731de24315eddedf/compiler/core/js_exp_make.ml#L1710-L1712)
>   ```ocaml
>   match (a.expression_desc, b.expression_desc) with
>   | ( (Null | Undefined _),
>      (Bool _ | Number _ | Typeof _ | Fun _ | Array _ | Caml_block _) ) ->
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   the `Typeof` or `Array` or `Caml_block` could have side effects, so can't always be safely dropped
> * in [`int_comp`](https://github.com/rescript-lang/rescript-compiler/blob/9b08fe929eb8aa06a0ab58f1731de24315eddedf/compiler/core/js_exp_make.ml#L1374-L1375)
>   ```ocaml
>     | Ceq, Optional_block _, Undefined _ | Ceq, Undefined _, Optional_block _ ->
>    false_
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   if `Optional_block` can have side effects then it shouldn't be dropped
> * in [`simplify_or`](https://github.com/rescript-lang/rescript-compiler/blob/9b08fe929eb8aa06a0ab58f1731de24315eddedf/compiler/core/js_exp_make.ml#L1071)
>   ```ocaml
>       | _, Bool true -> Some true_
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   the left operand shouldn't be dropped if it has side effects.
> 
> This is just a selection; there are some similar cases in those functions and others. I don't have all the development tools set up, so can't easily send a PR, unfortunately.

Thanks a lot @yallop !
I've fixed a bunch more in https://github.com/rescript-lang/rescript-compiler/pull/7151